### PR TITLE
[@types/react] Use ComponentType for React.Fragment

### LINKED ISSUE #21923: [@types/react]: React.Fragment does not have any construct or call signatures
- [x] I tried using the `@types/react` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @uniqueiniquity, @tkrotoff, @vsiao, 

There was some discussion about the proper type of `React.Fragment` in #21146, but while `symbol | number` might accurately reflect the internal React type for `Fragment`, it prevents `React.Fragment` from being used as a TSX element.

```tsx
<React.Fragment> // error here: React.Fragment does not have any construct or call signatures
  <div>Child 1</div>
  <div>Child 2</div>
</React.Fragment>
```

I realize we can render fragments using the `<></>` syntax, but the React docs state that this simply de-sugars to `<React.Fragment></React.Fragment>` and that [this is the only way to add keys to a fragment](https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html#keyed-fragments).

Therefore, I think the type of `React.Fragment` should be changed to something closer to `React.ComponentType<{}>`, which will give it `key` and `children` props.

I can open a PR if people agree with this proposal.

### PR DESCRIPTION
**Pull request by jgoz**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html#keyed-fragments

Fixes #21923 by using `ComponentType` as the type for `React.Fragment`.

Also:
- Updated existing `React.Fragment` tests
- Added TSX test for `React.Fragment` (but omitted `<></>` to preserve TS version compatibility)
- Fixed lint errors (`tslint:disable` is not allowed, prefer `disable-line` or `disable-next-line`)

### CODE REVIEW

**Comment by jgoz** [comment]
@uniqueiniquity As discussed in #21923. Please have a look.

File: types/react/index.d.ts [hunk 1]
```diff
@@ -281,10 +281,8 @@ declare namespace React {
         constructor(props: P, context?: any);
 
         // Disabling unified-signatures to have separate overloads. It's easier to understand this way.
-        // tslint:disable:unified-signatures
-        setState<K extends keyof S>(f: (prevState: Readonly<S>, props: P) => Pick<S, K>, callback?: () => any): void;
-        setState<K extends keyof S>(state: Pick<S, K>, callback?: () => any): void;
-        // tslint:enable:unified-signatures
+        setState<K extends keyof S>(f: (prevState: Readonly<S>, props: P) => Pick<S, K>, callback?: () => any): void; // tslint:disable-line:unified-signatures
```
Comment (vsiao) [inline types/react/index.d.ts]: can we make the tslint comments `disable-next-line` instead of adding them to the same line?

**Review by vsiao — APPROVED** [review]

File: types/react/index.d.ts (hunk 1 above)
Comment (jgoz) [reply types/react/index.d.ts]: Done.

**Comment by uniqueiniquity** [comment]
Can you also undo my changes on lines 91 and 225 [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21146/files#diff-96b72df8b13a8a590e4f160cbc51f40c)?

**Comment by jgoz** [comment]
@uniqueiniquity Done.

**Comment by uniqueiniquity** [comment]
Thank you, @jgoz!

**Comment by oknoorap** [comment]
is `@types/react` fragment fix already published on npm?

**Comment by uniqueiniquity** [comment]
@oknoorap yes, it is.

**Comment by oknoorap** [comment]
I've been use it 3 days ago, still not fixed, is this fixed right now?

**Comment by jgoz** [comment]
This has been fixed for about a week. Make sure you update to the latest version of `@types/react`, including transitive dependencies.

**Pull request merged**